File: typescript/abstract_factory/furniture/product.ts
export interface Chair {
  chair(): string
}

export class ArtDecoChair implements Chair {
  public chair(): string {
    const result = 'art deco chair'
    console.log(result)
    return result
  }
}

export class VictorianChair implements Chair {
  public chair(): string {
    const result = 'victorian chair'
    console.log(result)
    return result
  }
}

export class ModernChair implements Chair {
  public chair(): string {
    const result = 'modern chair'
    console.log(result)
    return result
  }
}

export interface Table {
  table(): string
}

export class ArtDecoTable implements Table {
  public table(): string {
    const result = 'art deco table'
    console.log(result)
    return result
  }
}

export class VictorianTable implements Table {
  public table(): string {
    const result = 'victorian table'
    console.log(result)
    return result
  }
}

export class ModernTable implements Table {
  public table(): string {
    const result = 'modern table'
    console.log(result)
    return result
  }
}

export interface Sofa {
  sofa(): string
}

export class ArtDecoSofa implements Sofa {
  public sofa(): string {
    const result = 'art deco sofa'
    console.log(result)
    return result
  }
}

export class VictorianSofa implements Sofa {
  public sofa(): string {
    const result = 'victorian sofa'
    console.log(result)
    return result
  }
}

export class ModernSofa implements  Sofa {
  public sofa(): string {
    const result = 'modern sofa'
    console.log(result)
    return result
  }
}

File: dart/creational/abstract_factory/furniture/bin/product.dart
abstract class Chair {
  String chair();
}

class ArtDecoChair implements Chair {
  @override
  String chair() {
    const result = 'art deco chair';
    print(result);
    return result;
  }
}

class VictorianChair implements Chair {
  @override
  String chair() {
    const result = 'victorian chair';
    print(result);
    return result;
  }
}

class ModernChair implements Chair {
  @override
  String chair() {
    const result = 'modern chair';
    print(result);
    return result;
  }
}

abstract class Table {
  String table();
}

class ArtDecoTable implements Table {
  @override
  String table() {
    const result = 'art deco table';
    print(result);
    return result;
  }
}

class VictorianTable implements Table {
  @override
  String table() {
    const result = 'victorian table';
    print(result);
    return result;
  }
}

class ModernTable implements Table {
  @override
  String table() {
    const result = 'modern table';
    print(result);
    return result;
  }
}

abstract class Sofa {
  String sofa();
}

class ArtDecoSofa implements Sofa {
  @override
  String sofa() {
    const result = 'art deco sofa';
    print(result);
    return result;
  }
}

class VictorianSofa implements Sofa {
  @override
  String sofa() {
    const result = 'victorian sofa';
    print(result);
    return result;
  }
}

class ModernSofa implements Sofa {
  @override
  String sofa() {
    const result = 'modern sofa';
    print(result);
    return result;
  }
}
